core/primitives/src/shard_layout.rs

```diff
@@ -83,17 +98,201 @@ pub struct ShardLayoutV1 {
+    }
+
+    pub fn new_end(start: AccountId) -> Self {
+        Self { start: AccountBoundary::Middle(start), end: AccountBoundary::End }
+    }
+
+    pub fn new_mid(start: AccountId, end: AccountId) -> Self {
+        Self { start: AccountBoundary::Middle(start), end: AccountBoundary::Middle(end) }
+    }
+
+    fn contains(&self, account_id: &AccountId) -> bool {
+        match (&self.start, &self.end) {
+            (AccountBoundary::Start, AccountBoundary::End) => true,
+            (AccountBoundary::Start, AccountBoundary::Middle(end)) => account_id < end,
+            (AccountBoundary::Middle(start), AccountBoundary::End) => account_id >= start,
+            (AccountBoundary::Middle(start), AccountBoundary::Middle(end)) => {
+                account_id >= start && account_id < end
+            }
+            (AccountBoundary::End, _) => panic!("invalid account range"),
+            (_, AccountBoundary::Start) => panic!("invalid account range"),
+        }
+    }
+}
+
+/// A mapping from the account range to shard id.
+/// e.g.
+///     (Start, "kkk") -> 0
+///     ("kkk", "ppp") -> 3
+///     ("ppp", End) -> 4
+type AccountRangeShardMap = BTreeMap<AccountRange, ShardId>;
+
+/// Validate the mapping from the account range to shard id.
+/// The validations consists of the following steps:
+/// - The first range must start with Start
+/// - The last range must end with End
+/// - All other boundaries must be Middle
+/// - There must not be any gaps or overlaps e.i.
+///     the end of previous range must equal to the start of the next one
+/// - The shard ids are unique
+fn validate_account_range_shard_mapping(
+    account_range_shard_mapping: &AccountRangeShardMap,
+) -> Result<(), ShardLayoutError> {
+    let err =
+        |reason: &str| ShardLayoutError::InvalidShardsAccountRange { reason: reason.to_string() };
+
+    let account_ranges = account_range_shard_mapping.keys().collect_vec();
+
+    // Check start and end.
+    let (Some(first), Some(last)) = (account_ranges.first(), account_ranges.last()) else {
+        return Err(err("account range empty"));
+    };
+    if first.start != AccountBoundary::Start {
+        return Err(err("first account range should start with Start"));
+    }
+    if last.end != AccountBoundary::End {
+        return Err(err("last account range should end with End"));
+    }
+
+    // Check middle ranges and no gaps or overlaps.
+    for i in 1..account_ranges.len() {
+        let prev = account_ranges[i - 1];
+        let curr = account_ranges[i];
+
+        if !matches!(prev.end, AccountBoundary::Middle(_)) {
+            return Err(err(&format!(
+                "Account range in the middle should not end with Start/End. {prev:?}"
+            )));
+        }
+        if !matches!(curr.start, AccountBoundary::Middle(_)) {
+            return Err(err(&format!(
+                "Account range in the middle should not start with Start/End. {curr:?}"
```

Comment:
nit: this could fire also for last account range, not only middle ones. Basically any non first account range

Reply:
I believe it does, in the loop the `prev` goes over [0, n-1) and `curr` goes over [1,n). The checks validate `prev.end` and `curr.start`, that should be the entire set of non-edge boundaries. Please correct me if I'm wrong.

Reply:
Correct
I was just reading the wording "middle" as the shard in the middle, but here it's used to indicate the type of AccountBoundary.. so all good 😃